### PR TITLE
#2245 Splitting LFRicLoopBounds from dynamo0p3.py

### LINKED ISSUE #2245: Splitting LFRicLoopBounds from `src/psyclone/dynamo0p3.py`
Part of the splitting dynamo0p3.py work detailed in #2235 

LFRicLoopBounds appears in `domain/lfric/lfric_invoke.py` and `tests/domain/lfric/lfric_loop_bounds_test.py`, no name change needed - uses LFRicCollection, so unlikely to have a circular dependency

Approximate order of things that need to be done:
* create new file `domain/lfric/lfric_loop_bounds.py`, add the license to top (and your name to modifiers!*)
* move LFRicLoopsBounds + imports across from `dynamo0p3.py` (don't forget documentation util at the bottom!)
* `pylint dynamo0p3.py`, remove unused imports
* `pylint domain/lfric/lfric_loop_bounds.py`, ditto
* `pycodestyle` the same
* add `lfric_loop_bounds.py` to `domain/lfric/__init__.py`
* change location of import to `from psyclone.domain.lfric import LFRicLoopBounds` in the files it is imported in - remove any `disable=import-outside-top-level` that are no longer used
* run `pylint` and `pycodestyle` on all files changed
* run `pytest`

*this is for all files you change, check the year range of the copyright too and update if needed

### PR DESCRIPTION
**Pull request by oakleybrunt**
closes #2245

### CODE REVIEW

**Comment by oakleybrunt** [comment]
Added more imported functions to the `__all__` attributes in `__init__.py`

**Comment by mo-lottieturner** [comment]
#2235

File: src/psyclone/domain/lfric/lfric_loop_bounds.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+''' This module implements the PSyclone Dynamo 0.3 API by 1)
+    specialising the required base classes in parser.py (KernelType) and
+    adding a new class (DynFuncDescriptor03) to capture function descriptor
+    metadata and 2) specialising the required base classes in psyGen.py
+    (PSy, Invokes, Invoke, InvokeSchedule, Loop, Kern, Inf, Arguments and
+    Argument). '''
+
+# Imports
+from psyclone.configuration import Config
+from psyclone.domain.lfric import (LFRicCollection)
```
Comment (arporter) [inline src/psyclone/domain/lfric/lfric_loop_bounds.py:48]: No need for parentheses here or below.

File: src/psyclone/tests/domain/lfric/lfric_loop_bounds_test.py [hunk 2]
```diff
@@ -32,15 +32,15 @@
 # POSSIBILITY OF SUCH DAMAGE.
 # -----------------------------------------------------------------------------
 # Author: A. R. Porter, STFC Daresbury Lab
-# Modified: L. Turner, Met Office
+# Modified: L. Turner and O. Brunt, Met Office
 
 ''' This module contains pytest tests for the LFRicLoopBounds collection
 class. '''
 
 from __future__ import absolute_import
```
Comment (arporter) [inline src/psyclone/tests/domain/lfric/lfric_loop_bounds_test.py:40]: Please rm this line (it was needed back when we supported both Python 2 and Python 3 and we're gradually removing them now).

File: src/psyclone/domain/lfric/__init__.py [hunk 3]
```diff
@@ -63,18 +63,25 @@
...
+    'KernCallInvokeArgList',
     'KernelInterface',
     'KernStubArgList',
     'LFRicArgDescriptor',
     'LFRicCollection',
     'LFRicConstants',
     'LFRicExtractDriverCreator',
     'LFRicInvoke',
-    'LFRicSymbolTable']
+    'LFRicKernCallFactory',
```
[the remark this answers is not in the document]
Comment (oakleybrunt) [reply src/psyclone/domain/lfric/__init__.py:83]: Ah, sorry about the confusion! I was going through this first issue with Iva and we came across the __init__.py file which was missing several package names in this section. She said to add them but I see now that this is confusing and should be separated into its own issue. I'll revert these changes and only add LFRicLoopBounds

**Comment by arporter** [comment]
Hi @oakleybrunt, is this ready for review again now?

**Comment by oakleybrunt** [comment]
@arporter Apologies. Yes, it is ready for review. I've added the label and won't forget on the next PR!

File: src/psyclone/domain/lfric/lfric_loop_bounds.py [hunk 4]
```diff
@@ -36,25 +36,17 @@
 # Modified J. Henrichs, Bureau of Meteorology
 # Modified A. B. G. Chalk and N. Nobre, STFC Daresbury Lab
 
-''' This module implements the PSyclone Dynamo 0.3 API by 1)
-    specialising the required base classes in parser.py (KernelType) and
-    adding a new class (DynFuncDescriptor03) to capture function descriptor
-    metadata and 2) specialising the required base classes in psyGen.py
-    (PSy, Invokes, Invoke, InvokeSchedule, Loop, Kern, Inf, Arguments and
-    Argument). '''
+''' This module Handles all variables required for specifying loop limits within a
```
Comment (arporter) [inline src/psyclone/domain/lfric/lfric_loop_bounds.py:39]: Please reword to say it provides the `LFRicLoopBounds` class that .... for the LFRic PSy layer.

File: src/psyclone/domain/lfric/lfric_loop_bounds.py [hunk 5]
```diff
@@ -36,25 +36,17 @@
...
 
 # Imports
 from psyclone.configuration import Config
-from psyclone.domain.lfric import (LFRicCollection)
-from psyclone.f2pygen import (AssignGen, CommentGen, DeclGen)
+from psyclone.domain.lfric import LFRicCollection
+from psyclone.f2pygen import AssignGen, CommentGen, DeclGen
 
 
 class LFRicLoopBounds(LFRicCollection):
```
Comment (arporter) [inline src/psyclone/domain/lfric/lfric_loop_bounds.py:48]: Please put back this doc string - all classes must have one.

**Review by arporter — CHANGES_REQUESTED** [review]
Almost there now.
Docstrings need a bit more work.
I've triggered the integration tests so will check those next time.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Ref. guide builds fine.
Integration tests are fine.
Will proceed to merge.

**Pull request merged**